File: kernel/processes/manager.lua
---@class ProcessManager
---@field process_id number
---@field processes table<number, Process>
---@field active_process Process
---@field queues table<number, Process[]>
---@field meta table<number, { queue_level: number, ticks_used: number }>
---@field tick_count number
ProcessManager = {}

-- MLFQ configuration
local NUM_QUEUES     = 3
local QUANTA         = {2, 4, 8}  -- yields per queue before demotion
local BOOST_INTERVAL = 50         -- steps between starvation-prevention boosts

function ProcessManager:new(arch)
    local procman = {
        process_id     = 0,
        processes      = {},
        active_process = nil,
        -- queues[1] = highest priority, queues[NUM_QUEUES] = lowest
        queues         = {{}, {}, {}},
        -- pid -> { queue_level (0-based), ticks_used }
        meta           = {},
        tick_count     = 0,
    }
    setmetatable(procman, self)
    self.__index = self
    return procman
end

function ProcessManager:add_process(process)
    process.pid     = self.process_id
    self.process_id = self.process_id + 1
    self.processes[process.pid] = process

    self.meta[process.pid] = {
        queue_level = 0,
        ticks_used  = 0,
    }
    table.insert(self.queues[1], process)
end

function ProcessManager:remove_process(pid)
    if self.processes[pid] then
        self.processes[pid].dead = true
    end
    self.processes[pid] = nil
    self.meta[pid]      = nil
end

-- Dequeue the next runnable process from the highest non-empty queue.
function ProcessManager:_next_process()
    for level = 1, NUM_QUEUES do
        local queue = self.queues[level]
        while #queue > 0 do
            local proc = table.remove(queue, 1)
            if not proc.dead then
                return proc, level - 1  -- 0-based level
            end
        end
    end
    return nil, nil
end

-- Move all processes into queue 0 to prevent starvation.
function ProcessManager:_boost_all()
    for level = 2, NUM_QUEUES do
        for _, proc in ipairs(self.queues[level]) do
            if not proc.dead then
                local m = self.meta[proc.pid]
                if m then
                    m.queue_level = 0
                    m.ticks_used  = 0
                end
                table.insert(self.queues[1], proc)
            end
        end
        self.queues[level] = {}
    end
end

-- Run one scheduler step: resume the next highest-priority process.
-- Returns true if a process was run, false if all queues are empty.
function ProcessManager:step()
    self.tick_count = self.tick_count + 1

    if self.tick_count % BOOST_INTERVAL == 0 then
        self:_boost_all()
    end

    local proc, level = self:_next_process()
    if not proc then return false end

    self.active_process = proc
    local m = self.meta[proc.pid]

    local ok = coroutine.resume(proc.coroutine)

    self.active_process = nil

    if not ok then
        proc.dead      = true
        proc.exit_code = -1
        self.meta[proc.pid] = nil
        return true
    end
    if coroutine.status(proc.coroutine) == "dead" then
        proc.dead      = true
        proc.exit_code = proc.exit_code or 0
        self.meta[proc.pid] = nil
        return true
    end

    if not m then return true end  -- process removed itself during execution
    m.ticks_used = m.ticks_used + 1

    if m.ticks_used >= QUANTA[level + 1] then
        -- Exhausted quantum, demote to next queue.
        m.ticks_used  = 0
        local new_level = math.min(level + 1, NUM_QUEUES - 1)
        m.queue_level = new_level
        table.insert(self.queues[new_level + 1], proc)
    else
        -- Remaining ticks in quantum; re-enqueue at same level.
        table.insert(self.queues[level + 1], proc)
    end

    return true
end

File: kernel/processes/manager.ts
import * as Process from "processes.process";

let NUM_QUEUES: integer = 3;
let QUANTA: integer[] = [2, 4, 8]; // slices per queue before demotion
let BOOST_INTERVAL = 50; // steps between starvation-prevention boosts
let INSTRUCTION_QUOTA = 500; // vm instructions per slice in preemptive mode

type MetaInfo = { queue_level: integer, ticks_used: integer }

export class ProcessManager {
	process_id: integer = 1;
	processes: Process[] = [];
	active_process: Process | null = null;
	queues: Process[][] = [];
	meta: MetaInfo[] = [];
	tick_count: integer = 0;
	preemptive: boolean = this.detect_preemption();

	constructor(arch: unknown) {
		for (let i = 0; i < NUM_QUEUES; i++) {
			this.queues.push([]);
		}
	}

	private detect_preemption() {
		let co = coroutine.create(function () { while (true) { } });
		debug.sethook(co, coroutine.yield, "", 50);
		let ok = coroutine.resume(co);
		// @ts-expect-error the library type for coroutines appears to be wrong
		debug.sethook(co, null);
		return ok && coroutine.status(co) == "suspended";
	}

	add_process(process: Process) {
		process.pid = this.process_id;
		this.process_id += 1;
		this.processes[process.pid] = process;

		this.meta[process.pid] = {
			queue_level: 0,
			ticks_used: 0
		};

		this.queues[0].push(process);
	}

	get_process(pid: integer): Process | null {
		return this.processes[pid] ?? null;
	}

	remove_process(pid: integer) {
		if (this.processes[pid]) {
			this.processes[pid].dead = true
		}

		delete this.processes[pid];
		delete this.meta[pid];
	}

	private next_process(): LuaMultiReturn<[Process, integer] | [null, null]> {
		for (let level = 0; level < NUM_QUEUES; level++) {
			let queue = this.queues[level]
			while (queue.length > 0) {
				let process = queue.shift();
				if (process !== undefined && !process.dead) {
					return $multi(process, level);
				}
			}
		}
		return $multi(null, null);
	}

	private boost_all() {
		for (let level = 1; level < NUM_QUEUES; level++) {
			for (const process of this.queues[level]) {
				if (!process.dead) {
					let meta = this.meta[process.pid];
					if (meta !== null) {
						meta.queue_level = 0;
						meta.ticks_used = 0;
					}
					this.queues[0].push(process);
				}
			}
			this.queues[level] = [];
		}
	}

	step(): boolean {
		this.tick_count += 1;

		if (this.tick_count % BOOST_INTERVAL === 0) {
			this.boost_all();
		}

		let [process, level] = this.next_process();
		if (process === null || level === null) { return false; }

		this.active_process = process;
		let meta = this.meta[process.pid];

		if (this.preemptive) {
			debug.sethook(process.coroutine, coroutine.yield, "", INSTRUCTION_QUOTA);
		}

		let [ok, err] = coroutine.resume(process.coroutine);

		if (this.preemptive) {
			// @ts-expect-error the library type for coroutines appears to be wrong
			debug.sethook(process.coroutine, null);
		}

		this.active_process = null;

		if (!ok) {
			let stderr = process.fds[2];
			if (stderr !== null) {
				stderr.write(tostring(err) + "\n");
			}
			process.dead = true;
			process.exit_code = -1;
			delete this.meta[process.pid];
			return true;
		}

		if (coroutine.status(process.coroutine) === "dead") {
			process.dead = true;
			process.exit_code = process.exit_code || 0;
			delete this.meta[process.pid];
			return true;
		}

		if (meta === null) { return true; }
		meta.ticks_used += 1;

		if (meta.ticks_used >= QUANTA[level]) {
			meta.ticks_used = 0;

			let new_level = math.min(level + 1, NUM_QUEUES - 1);
			meta.queue_level = new_level;
			this.queues[new_level].push(process);
		} else {
			this.queues[level].push(process);
		}

		return true;
	}
}

export function create(arch: unknown): ProcessManager {
	return new ProcessManager(arch);
}
